### ISSUE TITLE
[no-invalid-this] False positives in properties and certain functions

### ISSUE THREAD

**Issue by josephzidell** [typescript-eslint/typescript-eslint#1980]
**Repro**

```JS
{
  "rules": {
    'no-invalid-this': 'off', // disable the base rule
    '@typescript-eslint/no-invalid-this': ['error'],
  }
}
```

```TS
class HomeController {
	// other properties omitted...

	private testVersions = {
		vtest: 'test.yml',
	};
	private liveVersions = {
		v0: 'v0.yml',
	};
	private versions = {...this.liveVersions, ...this.testVersions}; // <-- errors reported here

	// other functions omitted ...

	getRulesHandler = (req: Request, res: Response, next: NextFunction): express.Response => {
		this.versions[...]; // <-- errors reported here
	}
}
```

**Expected Result**
No errors

**Actual Result**
```
error  Unexpected 'this'  @typescript-eslint/no-invalid-this
```

**Additional Info**
This is a controller for an [express.js](http://expressjs.com/) app

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^2.31.0` |
| `@typescript-eslint/parser`        | `^2.31.0` |
| `TypeScript`                       | `^3.8.3` |
| `ESLint`                           | `^6.8.0` |
| `node`                             | `v13.8.0` |
| `npm`                              | `6.14.5` |

**Comment by bradzacher** [comment]
duplicate of #491

**Comment by josephzidell** [comment]
@bradzacher There are two false positives listed here, and #491 does not cover this:

```ts
getRulesHandler = (req: Request, res: Response, next: NextFunction): express.Response => {
	this.versions[...]; // <-- errors reported here
}
```

**Comment by bradzacher** [comment]
They are the same false positive.
the rule doesn't know that class properties are a thing, so it doesn't know that a class declaration body provides an implicit `this` reference.

Which means that:
- it doesn't know that there is a `this` reference that class properties are able to interact with.
- it doesn't know that there is a `this` reference that an arrow function expression can close over.

same bug - once the rule is taught that there's a `this` reference implied by the class body, then both issues are fixed.
